src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue`1.cs

```diff
@@ -38,11 +39,44 @@ internal class AsyncBatchingWorkQueue<TItem> : AsyncBatchingWorkQueue<TItem, Voi
         }
 
         private static Func<ImmutableSegmentedList<TItem>, CancellationToken, ValueTask<VoidResult>> Convert(Func<ImmutableSegmentedList<TItem>, CancellationToken, ValueTask> processBatchAsync)
-            => async (items, ct) =>
+        {
+            return (items, cancellationToken) =>
             {
-                await processBatchAsync(items, ct).ConfigureAwait(false);
-                return default;
+                // PERF: for the fast path case where processBatchAsync runs to completion synchronously, this call to
+                // AsTask will simply return Task.CompletedTask.
+                var result = processBatchAsync(items, cancellationToken).AsTask();
+                if (TryWrapWithVoidResult(result) is { } wrapped)
+                    return new ValueTask<VoidResult>(wrapped);
+
+                return new ValueTask<VoidResult>(result.ContinueWith(
+                    static task => TryWrapWithVoidResult(task) ?? throw ExceptionUtilities.Unreachable(),
+                    CancellationToken.None,
+                    TaskContinuationOptions.LazyCancellation | TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default).Unwrap());
+
+                static Task<VoidResult>? TryWrapWithVoidResult(Task task)
+                {
+                    switch (task.Status)
+                    {
+                        case TaskStatus.RanToCompletion:
+                            return SpecializedTasks.Default<VoidResult>();
+
+                        case TaskStatus.Faulted:
+                            return Task.FromException<VoidResult>(task.Exception!.InnerException!);
+
+                        case TaskStatus.Canceled when task is { Exception.InnerException: OperationCanceledException { CancellationToken: var reportedCancellationToken } }:
+                            return Task.FromCanceled<VoidResult>(reportedCancellationToken);
+
+                        case TaskStatus.Canceled:
+                            Debug.Fail("This is not expected. Cancelled tasks are expected to provide a cancellation token.");
+                            return Task.FromCanceled<VoidResult>(new CancellationToken(canceled: true));
+
+                        default:
+                            return null;
+                    }
+                }
             };
```

Comment:
this logic is *extremely* difficult to rationalize.  i would far prefer to just wait for an updated compiler.

Reply:
I generally would agree, but this method is single-handedly responsible for over 109000 exceptions in 85 seconds (throw rate 1000 exceptions/second).

Reply:
but it looks like julien is literally in the middle of that PR, and is getting ready for it to be tested.  If they were thinking this was going to be coming out months from now, fine.  But we're literally overlapping the core work that seems to address this problem space is a healthier fashion.  

Note: we've had this design/issue for a while.  Waiting a short while for the compiler fix seems fine.

--

That said, i'm ok with us doing the work (and waiting on checking it in).  That way if the compiler work falls through, we have this in our pocket.  But, in a very real sense, this issue here serves as a great test for Julien's work.  It provides a severe, measurable, problem that we can compare new results against directly.